### ISSUE TITLE
Is it possible to add slurm rest API to a 3.0.0 pcluster?

### ISSUE THREAD

**Issue by c200chromebook** [aws/aws-parallelcluster#3492]
I know this is a little bit outside of the normal ticket categories, but is it possible to add slurm rest API to a pcluster install? When I propose `yum install slurm-slurmrestd` I get a plan that proposes installing slurm, which is, I don't think, what I want.

```

================================================================================
 Package                     Arch     Version                Repository    Size
================================================================================
Installing:
 slurm-slurmrestd            x86_64   20.11.8-2.el7          epel         132 k
Installing for dependencies:
 compat-gcc-48-libgfortran   x86_64   4.8.5-16.amzn2.0.4     amzn2-core   297 k
 freeipmi                    x86_64   1.5.7-3.amzn2          amzn2-core   2.0 M
 hdf5                        x86_64   1.8.12-13.el7          epel         1.6 M
 http-parser                 x86_64   2.7.1-8.amzn2.2        amzn2-core    30 k
 libaec                      x86_64   1.0.4-1.el7            epel          34 k
 libjwt                      x86_64   1.12.1-7.el7           epel          24 k
 munge                       x86_64   0.5.11-3.el7           epel          95 k
 munge-libs                  x86_64   0.5.11-3.el7           epel          37 k
 pmix                        x86_64   1.1.3-1.el7            epel         114 k
 slurm                       x86_64   20.11.8-2.el7          epel         1.7 M
 slurm-libs                  x86_64   20.11.8-2.el7          epel         1.0 M

Transaction Summary
================================================================================
Install  1 Package (+11 Dependent packages)

Total download size: 7.0 M
```

Yes, and indeed if I do that then restart the head node:
```
[ec2-user@ip-... gridwork]$ sinfo
slurm_load_partitions: Unable to contact slurm controller (connect failure)
```

Making some progress on this, does this require slurmdbd?

**Comment by c200chromebook** [comment]
Yes, and indeed if I do that then restart the head node:
```
[ec2-user@ip-10-45-114-103 gridwork]$ sinfo
slurm_load_partitions: Unable to contact slurm controller (connect failure)
```

**Comment by chenwany** [comment]
Hi! In Pcluster 3.0.0, we currently not enable slurmrestd by default.
In the [recipe](https://github.com/aws/aws-parallelcluster-cookbook/blob/970e9f504ced6e64760e47afd536972bfc2eaf08/recipes/slurm_install.rb#L82) we use to install slurm in ParallelCluster, we run `./configure --prefix=/opt/slurm --with-pmix=/opt/pmix`. By adding `--enable-slurmrestd`, it may be possible to enable slurmrestd.
`json-c-devel` and `http-parser-devel` are the dependencies required for using slurmrestd, you will need to manually install it.

Here's the step that I have to enable slurmrestd in 3.0.0. You could try if it works.
```
source /opt/parallelcluster/pyenv/versions/3.7.10/envs/node_virtualenv/bin/activate
cd /etc/chef/local-mode-cache/cache/slurm-slurm-20-11-8-1
yum install json-c-devel
yum install http-parser-devel
./configure --prefix=/opt/slurm --with-pmix=/opt/pmix --enable-slurmrestd
```
See if that works for you.

For the error you saw, you could check /var/log/slurmctld.log for details.

ParallelCluster currently doesn't support headnode restart, it will potentially cause some problems, please don't restart headnode, thank you!

**Comment by c200chromebook** [comment]
Thank you - can you comment as to what happens if you restart the head node? Would like to know what to look out for if it inadvertently goes down and comes back.

**Comment by jokramer123** [comment]
I have the same issue as the OP, when I manually install slurmrestd it will override the existing parallelcluster slurm installation.

I tried running the following as suggested by @chenwany and it doesn't work either:
```
source /opt/parallelcluster/pyenv/versions/3.7.10/envs/node_virtualenv/bin/activate
cd /etc/chef/local-mode-cache/cache/slurm-slurm-20-11-8-1
yum install json-c-devel
yum install http-parser-devel
./configure --prefix=/opt/slurm --with-pmix=/opt/pmix --enable-slurmrestd
```

Error:
```
slurmrestd
bash: slurmrestd: command not found...
```

Here is the OS and AMI I am using:

```
cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

curl http://169.254.169.254/latest/meta-data/ami-id
ami-0348652b6078c2c1d
```

Any suggestions?

**Comment by demartinofra** [comment]
Hi @jokramer123 first let me try to clarify what we provide out of the box with ParallelCluster 3.0 and then help you with this customization.

As @chenwany pointed out, the version of Slurm compiled and shipped with ParallelCluster 3.0 is not built with slurmrestd support. We are planning to compile Slurm with slurmrestd support as part of ParallelCluster 3.1 where we will potentially also upgrade to Slurm 21.08. However this does not mean that slurmrestd will be enabled or started out of the box.

In order to use slurmrestd as part of ParallelCluster 3.0 you'd have to recompile Slurm after installing jsonc and http-parser libraries that are required by slurmrestd. You can either recompile Slurm after you create your cluster or you can do this as part of a custom AMI creation. The instructions to recompile Slurm in a running cluster are the following:

1. Download the Slurm version you plan to use (we recommend to stick to the one shipped with the ParallelCluster version you are using). Note that ParallelCluster 3.0 is not compatible with Slurm 21.08.

```
curl https://download.schedmd.com/slurm/slurm-20.11.8.tar.bz2 -o /tmp/slurm-20.11.8.tar.bz2
tar -xf /tmp/slurm-20.11.8.tar.bz2  -C /tmp/
```
If you want you could reuse the locally cached version under `/etc/chef/local-mode-cache/cache/slurm-slurm-20-11-8-1`

2. Make sure there are no running compute nodes. If there are stop them with `pcluster update-compute-fleet --status STOP_REQUESTED --cluster-name name`

3. Stop slurmctld
```
sudo systemctl stop slurmctld
```

4. Install deps and recompile Slurm
```
cd /tmp/slurm-20.11.8/
sudo su -
yum install -y json-c-devel http-parser-devel

source /opt/parallelcluster/pyenv/versions/3.7.10/envs/cookbook_virtualenv/bin/activate
./configure --prefix=/opt/slurm --with-pmix=/opt/pmix --enable-slurmrestd
CORES=$(grep processor /proc/cpuinfo | wc -l)
make -j $CORES
make install
make install-contrib
deactivate
```

5. Start slurmctld
```
sudo systemctl start slurmctld
```

6. Follow SchedMD docs to enable and configure slurmrestd. slurmrestd binary should become available under `/opt/slurm/sbin/`.

Let me know if you need additional help or have additional questions.

**Comment by enrico-usai** [comment]
Content extended and moved into a wiki page: [How to enable slurmrestd on ParallelCluster 3.0](https://github.com/aws/aws-parallelcluster/wiki/How-to-enable-slurmrestd-on-ParallelCluster-3.0)
